Infer @PointerBounds macro from clang __counted_by parameters

hnrklssn opened the pull request:
This results in an automatic wrapper function with safe pointer types
when the imported function has bounds attributes. This exercises similar
pathways as the recently added functionality for specifying macros from
swift_attr, and fixes some bugs related to macro source file management.

Includes some unmerged commits from #76969.

Xazax-hun commented on CMakeLists.txt:
```diff
@@ -747,6 +747,10 @@ option(SWIFT_ENABLE_EXPERIMENTAL_PARSER_VALIDATION
   "Enable experimental SwiftParser validation by default"
   FALSE)
 
+option(SWIFT_ENABLE_EXPERIMENTAL_POINTER_BOUNDS
```
What is the main motivation to make this a compile time flag rather than just a runtime switch for the compiler?

Xazax-hun replied:
Any reason we want to keep this out of stdlib by default? My main concern is that we are not really testing this in the CI if this is off by default (unless we make this on in the CI).

hnrklssn replied:
I mostly followed what other experimental features did, so I'm not married to the approach. But I do have a concern that we'll want to make breaking changes to the macro API, so we need some way of making it clear that users shouldn't use this macro directly and expect source compatibility.
Note that this particular change is from https://github.com/swiftlang/swift/pull/76969, so we might want to take the discussion there.

Xazax-hun replied:
Ah, I see. So Swift has no mechanism equivalent to "unstable" features and APIs in Rust that can break any time? Now I better understand the motivation and if that is the case (no alternative), I am OK with this approach.

hnrklssn replied:
I'm not sure tbh, I'm completely new to swift. @DougGregor could you fill us in?

hnrklssn replied:
Right, so implicitly applying the macro on import could be an unstable language feature, but there's no way to include the macro in the stdlib in an "unstable" way, meaning that has to be stabilised before the importing?

Xazax-hun commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -8437,6 +8438,107 @@ ClangImporter::Implementation::importSwiftAttrAttributes(Decl *MappedDecl) {
+class PointerParamInfo {
+public:
+  virtual void print(clang::ASTContext &ctx, llvm::raw_ostream &out) const = 0;
+  virtual ~PointerParamInfo() {}
```
Nit: I think usually `= default;` is preferred over `{}`. Although I do not expect a difference in this particular case.

Xazax-hun commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -8437,6 +8438,107 @@ ClangImporter::Implementation::importSwiftAttrAttributes(Decl *MappedDecl) {
+  virtual ~PointerParamInfo() {}
+};
+
+class CountedByParam : public PointerParamInfo {
```
I presume `Clang` also has a representation of this information. Any reason we cannot reuse Clang's classes for that instead of rolling our own? E.g., we could just reuse `CountAttributedType` and add a custom print function for it.

hnrklssn replied:
We could do that, but we would need to make it a pair with both `CountAttributedType` and pointer index. When we add `DynamicRangeType` we also need the pointer index for the end pointer. Another option would be to not build a `SmallVector` at all, and print on the first pass. Not sure whether initialising `MacroString` and the outputstream when there are no bounds annotations to process is preferable. I guess if it turns out to be a problem we could mark functions for later bounds importing when we initially process parameters and encounter a bounds attribute?

hnrklssn replied:
Added a commit with another approach, please let me know what you think

Xazax-hun commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -8437,6 +8438,107 @@ ClangImporter::Implementation::importSwiftAttrAttributes(Decl *MappedDecl) {
+    FuncDecl *MappedDecl) {
+  auto ClangDecl =
+      dyn_cast_or_null<clang::FunctionDecl>(MappedDecl->getClangDecl());
+  if (!ClangDecl)
```
Could this be an assert? Do we ever expect this function to be invoked on something that does not have a clang declaration or the clang declaration is not a function declaration?

hnrklssn replied:
I found this comment in another function:
```
  // sourceDecl->getClangDecl() can be null because some lazily instantiated cases like C++ members that were instantiated from using-shadow-decls have no corresponding Clang decl.
```
My main motivation was that it is how `importSwiftAttrAttributes()` does it however, and the same argument is passed to both functions.

hnrklssn commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -574,6 +574,10 @@ void importer::getNormalInvocationArguments(
   }
 
+#ifdef SWIFT_ENABLE_EXPERIMENTAL_POINTER_BOUNDS
+  invokationsArgStrs.push_back("-fexperimental-bounds-safety-attributes");
```
This is going to be part of the bounds safety upstreaming effort

DougGregor replied:
Ah, okay. So that's an llvm-project change that allows the use of the `counted_by`, `sized_by`, etc. Clang attributes in headers outside of `-fbounds-safety`, right?

hnrklssn replied:
Indeed!

hnrklssn commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -2424,6 +2424,13 @@ ClangImporter::Implementation::importParameterType(
+    paramTy = CAT->desugar();
+    if (auto FuncD =
+            dyn_cast<clang::FunctionDecl>(param->getParentFunctionOrMethod()))
+      funcsWithPointerBounds.insert(FuncD);
```
Perhaps, I'd have to take a look at it. Where do I find that? The reason I'm not certain that's helpful for this case is that we still want to process an entire function decl and its parameters + return type all at once, to construct the singular `@PointerBounds` macro invocation. So the advantage here is that we know it's a parameter and how to get the parent function.

DougGregor replied:
I'm thinking about `ClangImporter::Implementation::importType`, which is what recursively walks a Clang type to transform it. The result is an `ImportedType`, which can carry additional information about things it saw along the way. For example, the `isIUO` bit is for places where we should have an implicitly-unwrapped optional, which is similar to the "has pointer bounds" info you are capturing here. Making this something that `ImportedType` tracks will make it easier to check that bit for (e.g.) all of the parameter and result types of a function or ObjC method, property, etc.

hnrklssn replied:
I'll take a look

hnrklssn replied:
I tried putting it in `ImportedType`, but it ended up not being useful, as nested pointers are imported using `importTypeIgnoreIUO` which throws away the `ImportedType`. While we don't support nested pointers right now, it seems like something worth doing the right way from the beginning.

hnrklssn commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -8437,6 +8438,100 @@ ClangImporter::Implementation::importSwiftAttrAttributes(Decl *MappedDecl) {
+  SourceLoc atLoc = parser.consumeToken(tok::at_sign);
+  DeclContext *DC = MappedDecl->getParent();
+  auto initContext = PatternBindingInitializer::create(DC);
+  hadError = parser
+                 .parseDeclAttribute(MappedDecl->getAttrs(), atLoc, atEndLoc,
+                                     initContext,
+                                     /*isFromClangAttribute=*/true)
+                 .isError();
```
As in generate a swift_attr and attach it to the decl, or break out the parsing to a common function?

DougGregor replied:
We have the notion that a `swift_attr`'s contents go into a Swift source file so the Swift parser can process it. That's how https://github.com/swiftlang/swift/pull/76364 makes `swift_attr` work, for example. I'd like to leverage that existing code.

hnrklssn replied:
That's what I mean. The code snippet you commented is copied from `importSwiftAttrAttributes()`. So to reuse it as shared code, the options would be to either process bounds attributes first and attach the result as a swift_attr, or extract the duplicated code into a helper function.

DougGregor commented:
With luck, https://github.com/swiftlang/swift/pull/77580 should make the rebased version of this PR functional. I'm sure I missed something there, though.

hnrklssn commented:
> With luck, #77580 should make the rebased version of this PR functional. I'm sure I missed something there, though.

Indeed, it works now! I'll start addressing the rest of the comments.

hnrklssn commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -8437,6 +8438,100 @@ ClangImporter::Implementation::importSwiftAttrAttributes(Decl *MappedDecl) {
+void ClangImporter::Implementation::importBoundsAttributes(
+    FuncDecl *MappedDecl) {
+  auto ClangDecl =
+      dyn_cast_or_null<clang::FunctionDecl>(MappedDecl->getClangDecl());
```
Running into some issue with missing SourceLocation during SILGen for methods, even after the `open`->`public final` transform, so punting ObjC to a follow-up.

DougGregor commented on include/swift/Basic/LangOptions.h:
```diff
@@ -325,6 +325,10 @@ namespace swift {
 
+    /// Enable C interop importing bounds safety and lifetime attributes to
+    /// generate Swift wrappers with safe pointer types (experimental).
+    bool EnableBoundsSafetyInterop = false;
```
I would prefer that we add a new experimental feature rather than a separate LangOption. I'm a little nervous about putting this under `LifetimeDependence`, because `LifetimeDependence` is going to get reviewed and enabled-by-default before we're ready to start performing this transform.

hnrklssn replied:
What is the difference between a language feature and and a flag?

hnrklssn commented on lib/Sema/TypeOfReference.cpp:
```diff
@@ -977,6 +977,12 @@ ConstraintSystem::getTypeOfReference(ValueDecl *value,
     return { openedType, openedType, openedType, openedType, Type() };
   }
 
+  // FIXME: implement this properly
```
@DougGregor I crashed here after rebasing onto your latest changes. The confident comment that VarDecl is the only remaining possibility makes me suspect this is not the right place to fix this however, so I didn't put much effort to figure out the Right Fix(tm).

DougGregor replied:
Okay. Do you need some help with something here, or was this just an FYI?

hnrklssn replied:
If there's anything that strikes you just from looking at it, about why we'd get this error for the special attribute source files but not otherwise, that'd be useful context. If not, I'll take some more time to investigate after the break.

hnrklssn replied:
I guess my point is, I suspect that the actual fix should be somewhere else, since EnumElements normally don't end up here. I just don't know where, yet.

hnrklssn replied:
The issue seems to be related to the cached parsing we do, because the crash only occurs when compiling a file with 2 identical macros, and only when those macros are applied by ClangImporter (so manually spelling the macro out in the source file doesn't trigger the crash).

When we resolve the first macro, we run typeCheckExpression on this expr:
```
(macro_expansion_expr type="<null>" name="PointerBounds" discriminator=65535
  (argument_list
    (argument
      (call_expr type="<null>" isolation_crossing="none"
        (unresolved_member_expr type="<null>" name="countedBy" function_ref=unapplied)
        (argument_list labels="pointer:count:"
          (argument label="pointer"
            (integer_literal_expr type="<null>" value="2" builtin_initializer="**NULL**" initializer="**NULL**"))
          (argument label="count"
            (string_literal_expr type="<null>" encoding=utf8 value="len" builtin_initializer="**NULL**" initializer="**NULL**")))))))
```. When we resolve the second one, the AST nodes are already resolved:
```
(macro_expansion_expr type="<null>" name="PointerBounds" discriminator=65535
  (argument_list
    (argument
      (call_expr type="PointerParam" location=:1:17 range=[:1:16 - line:1:51] isolation_crossing="none"
        (dot_syntax_call_expr type="(Int, String) -> PointerParam" location=:1:17 range=[:1:16 - line:1:17] isolation_crossing="none"
          (declref_expr type="(PointerParam.Type) -> (Int, String) -> PointerParam" location=:1:17 range=[:1:17 - line:1:17] decl="Swift.(file).PointerParam.countedBy(pointer:count:)" function_ref=single)
          (argument_list implicit
            (argument
              (type_expr implicit type="PointerParam.Type" location=:1:16 range=[:1:16 - line:1:16] typerepr="PointerParam"))))
        (argument_list labels="pointer:count:"
          (argument label="pointer"
            (integer_literal_expr type="Int" location=:1:36 range=[:1:36 - line:1:36] value="2" builtin_initializer="Swift.(file).Int.init(_builtinIntegerLiteral:)" initializer="**NULL**"))
          (argument label="count"
            (string_literal_expr type="String" location=:1:46 range=[:1:46 - line:1:46] encoding=utf8 value="len" builtin_initializer="Swift.(file).String extension.init(_builtinStringLiteral:utf8CodeUnitCount:isASCII:)" initializer="**NULL**")))))))
```
This infrastructure is not set up to handle resolved AST nodes like `dot_syntax_call_expr`, because it should always run before those are created. For some reason the result of the `ResolveMacroRequest` doesn't seem to be cached properly, despite seemingly having the same inputs. Looking into why that is.

DougGregor replied:
Ah, I understand this now. It's because we aren't properly cloning the expression within each CustomAttr node. I'll think about how to address this.

hnrklssn replied:
I tried doing the opposite, and not clone the CustomAttr node in the first place. At least at first glance, it seems to work and the ResolveMacroRequest is cached as it should. Would this be problematic for some reason I'm missing?

DougGregor replied:
I expect that means that you aren't getting the attribute applied?

DougGregor commented:
> What is the difference between a language feature and and a flag?

There's a bunch of machinery around language features and their lifecycle that we don't get from a one-off flag. Plus, flags need to be manually handled in the driver, whereas language features are already available.

hnrklssn commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -8532,6 +8544,78 @@ ClangImporter::Implementation::importSwiftAttrAttributes(Decl *MappedDecl) {
+
+    size_t parameterIndex = 1;
+    PointerParamInfoPrinter printer(getClangASTContext(), out);
+    for (auto param : ClangDecl->parameters()) {
```
interleaveComma ended up complicating things more than it simplified, because it requires first filtering out all the non-bounds safe parameters first to avoid introducing extraneous commas.

hnrklssn commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3391,3 +3396,62 @@ SourceLoc PrettyPrintDeclRequest::evaluate(Evaluator &eval, const Decl *decl) co
+  std::string bufferName;
+  {
+    llvm::raw_string_ostream out(bufferName);
+    for (auto iter = nameComponents.rbegin(); iter != nameComponents.rend();
```
Removed this change because it's no longer needed.

hnrklssn committed:
Infer @PointerBounds macro from clang __counted_by parameters

This results in an automatic wrapper function with safe pointer types
when the imported function has bounds attributes. This exercises similar
pathways as the recently added functionality for specifying macros from
swift_attr. The new functionality is guarded by the experimental
language feature SafeInteropWrappers.

rdar://97942270

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented:
After all the iterations of rebasing most of the intermediate commits didn't build anymore, and many contained a bunch of code that's removed in later commits, so I squashed them

DougGregor commented:
@swift-ci please smoke test Windows
